src/lib/logical_query_plan/join_node.cpp

```diff
@@ -228,19 +230,72 @@ bool JoinNode::is_column_nullable(const ColumnID column_id) const {
 
 const std::vector<std::shared_ptr<AbstractExpression>>& JoinNode::join_predicates() const { return node_expressions; }
 
+void JoinNode::mark_as_semi_reduction(const std::shared_ptr<JoinNode>& reduced_join_node) {
+  Assert(!_is_semi_reduction, "The semi reducer status should be set once only.");
+  Assert(reduced_join_node, "Reduced JoinNode must be provided.");
+  Assert(join_mode == JoinMode::Semi, "Semi join reductions require JoinMode::Semi.");
+  DebugAssert(join_predicates().size() == 1,
+              "Currently, semi join reductions are expected to have a single join predicate.");
+  DebugAssert(std::any_of(reduced_join_node->join_predicates().begin(), reduced_join_node->join_predicates().end(),
+                          [&](const auto predicate) { return *predicate == *join_predicates()[0]; }),
+              "Both semi join reduction node and the reduced join should have a common join predicate.");
+  _is_semi_reduction = true;
+  _reduced_join_node = std::weak_ptr<JoinNode>(reduced_join_node);
+}
+
+bool JoinNode::is_semi_reduction() const {
+  DebugAssert(!_is_semi_reduction || join_mode == JoinMode::Semi, "Non-semi join is marked as a semi reduction.");
+  return _is_semi_reduction;
+}
+
+std::shared_ptr<JoinNode> JoinNode::get_or_find_reduced_join_node() const {
+  if (!_is_semi_reduction) {
+    return nullptr;
+  }
+
+  if (_reduced_join_node.expired()) {
+    // In deep copies of the LQP, the weak pointer to the reduced join is expired (lazy discovery). In such cases,
+    // find the reduced join by traversing the LQP upwards.
+    visit_lqp_upwards(std::const_pointer_cast<AbstractLQPNode>(shared_from_this()), [&](const auto& current_node) {
+      if (current_node->type != LQPNodeType::Join || current_node.get() == this) {
```

Comment:
We can't use `current_node == /*current_node ==` because we do not take care of the reducer status in the `_on_shallow_equals` dunction, right? However, is there a nicer way to compare that the pointers poinmt to the same address?

Reply:
Not really. It is just a (cheap) check to skip the first node in the upwards traversal.

Reply:
> However, is there a nicer way to compare that the pointers poinmt to the same address?

I guess not, but happy to hear, if so. 😄